### PR TITLE
new feature: Wallet Receive Page

### LINKED ISSUE #965: Allow receiving funds on BTCPay Wallet
**Is your feature request related to a problem? Please describe.**
BTCPay is a very feature-rich wallet. However, there is no way outside of generating invoices(through apps or manuall) to be able to get an address to send bitcoin to directly.

**Describe the solution you'd like**
Add a new section to the wallet UI: "Receiving"
Here, you can click a button "Generate new address" which would reserve a new address to send bitcoin to and display it on screen with a QR code. Optionally, you should also be able to pre-assign labels to this receiving address so that once funds are received, the wallets shows the label.

**Describe alternatives you've considered**
Create an invoice manually or through an app. This has the drawback that the invoice shows a specific amount and an unnecessary UI for something you want quick. 

**Provide examples**
Transferring some bitcoin from an exchange to your cold-storage wallet linked to your btcpay store. Just generate a receiving address and attach a label "funds bought from binance exchange after receiving August salary "

### PR DESCRIPTION
**Pull request by Kukks**
Very simple for now but we can improve on UX and UI afterwards
image

closes #965

### CODE REVIEW

**Comment by NicolasDorier** [comment]
YES!
Can you make it the default page instead of /transactions though? Because for big wallet /transactions is slow to open.

File: BTCPayServer/Controllers/WalletsController.cs [hunk 1]
```diff
@@ -271,6 +271,26 @@ private static string GetLabelTarget(WalletId walletId, uint256 txId)
...
+            var store = await Repository.FindStore(walletId.StoreId, GetUserId());
+            DerivationSchemeSettings paymentMethod = GetDerivationSchemeSettings(walletId, store);
+            if (paymentMethod == null)
+                return NotFound();
+            var network = this.NetworkProvider.GetNetwork<BTCPayNetwork>(walletId?.CryptoCode);
+            if (network == null)
+                return NotFound();
+
+            var wallet = _walletProvider.GetWallet(network);
+            var address = await wallet.ReserveAddressAsync(paymentMethod.AccountDerivation);
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/WalletsController.cs]: Don't reserve everytimes.

**Comment by NicolasDorier** [comment]
So here are the difficult parts.

1. You should not generate a new address at each refresh. This is a GET and should not have side effect.
2. You can't just show the next unused address, because it is possible that an invoice comes in with this address, and the payment then get attributed to the invoice by mistake.

It means that either: 
1. You must show a button showing "Get a new address".
2. You must reserve the address the first time and show it until it get actually used. (I can maybe add something in NBXplorer to know this)

**Comment by Kukks** [comment]
Yes I ran into this dilemma and opted to still make the most basic version of this PR for the sake of discussion.
Coin logo is no issue, will add it in a bit.

I think it should be that
* browse to /receive
* Check if we have reserved address for wallet receive ( memory cache?)
* if we do  have a reserved address, check if it was used.
  * If used, clear cache and generate new address( and save to cache)
  * If not used, display address
* If we do not, generate new address( and save to cache)
*show address on screen + a button to "Generate a new address"

We will also need an even subscription to sore derivation changes so that we clear cache so that we dont show an incorrect address from cache

The main issue here is that you still generate a new address per btcpay run but it's not too bad

**Comment by pavlenex** [comment]
Oh boy, I wanted this so bad.

About the new unused address, I think, generate a new address button should be enough?

Bisq has an interesting UI for that, but perhaps it's a bit of an overkill, but the bottom part of the image can be handy for us?

bisq

**Comment by Kukks** [comment]
> 
> 
> Yes I ran into this dilemma and opted to still make the most basic version of this PR for the sake of discussion.
> Coin logo is no issue, will add it in a bit.
> 
> I think it should be that
> 
>     * browse to /receive
> 
>     * Check if we have reserved address for wallet receive ( memory cache?)
> 
>     * if we do  have a reserved address, check if it was used.
>       
>       * If used, clear cache and generate new address( and save to cache)
>       * If not used, display address
> 
>     * If we do not, generate new address( and save to cache)
>       *show address on screen + a button to "Generate a new address"
> 
> 
> We will also need an even subscription to sore derivation changes so that we clear cache so that we dont show an incorrect address from cache
> 
> The main issue here is that you still generate a new address per btcpay run but it's not too bad

Done!

**Comment by NicolasDorier** [comment]
> We will also need an even subscription to sore derivation changes so that we clear cache so that we dont show an incorrect address from cache

Or I maybe can just add an API at NBXplorer level to know about this.

**Comment by Kukks** [comment]
rebased

**Comment by NicolasDorier** [comment]
tests don't pass

**Comment by pavlenex** [comment]
Looks great. I've tested it, but is it just me or the wording `Unreserve BTC Address` is a bit odd? Perhaps `Skip`? I completely understand what it does, but not sure if there's a better wording that may even be easier to translate in the future.

**Comment by britttttk** [comment]
tACK - The `Unreserve this Address` text is clear to me.

**Comment by Kukks** [comment]
tests pass @NicolasDorier  
It can definitely be improved with some extra functionality from NBX but this preserves addresses as much as possible and should be a decent base

**Comment by pavlenex** [comment]
> tACK - The `Unreserve this Address` text is clear to me.

Dunno, I can't even think how `unreserve` can be translated. `Next Address` is easier, but if you guys are for the term, I'm fine with it. We'll see if it bothers translators, we're still far away from #972

**Comment by Kukks** [comment]
> 
> 
> > tACK - The `Unreserve this Address` text is clear to me.
> 
> Dunno, I can't even think how `unreserve` can be translated. `Next Address` is easier, but if you guys are for the term, I'm fine with it. We'll see if it bothers translators, we're still far away from #972

Next address is not the same as unreserve

Unreserve means that you can use this address in the next invoice if you click it
next address means that you leave the current reserved and reserve the next one too

**Comment by pavlenex** [comment]
> > > tACK - The `Unreserve this Address` text is clear to me.
> > 
> > 
> > Dunno, I can't even think how `unreserve` can be translated. `Next Address` is easier, but if you guys are for the term, I'm fine with it. We'll see if it bothers translators, we're still far away from #972
> 
> Next address is not the same as unreserve
> 
> Unreserve means that you can use this address in the next invoice if you click it
> next address means that you leave the current reserved and reserve the next one too

Wait, but the address changes upon receiving balance? Why would I want to `unreserve` the address, users don't care, if no balance show this address, when an invoice is generated with that address, show different one, I don't see why users even need to do anything, as it may encourage address re-use, no?

I'd probably have to take another look at the PR, I forgot how it works, but if there's a concensus on the term, I'm fine with it, perhaps it's just me that finds it unconventional (I've never seen a software or a wallet using the term) and difficult to translate.

File: BTCPayServer/Controllers/WalletsController.cs [hunk 2]
```diff
@@ -276,6 +286,68 @@ private static string GetLabelTarget(WalletId walletId, uint256 txId)
...
+            if (walletId?.StoreId == null)
+                return NotFound();
+            DerivationSchemeSettings paymentMethod = GetDerivationSchemeSettings(walletId);
+            if (paymentMethod == null)
+                return NotFound();
+            var network = this.NetworkProvider.GetNetwork<BTCPayNetwork>(walletId?.CryptoCode);
+            if (network == null)
+                return NotFound();
+
+            var address = _WalletReceiveStateService.Get(walletId)?.ScriptPubKey
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/WalletsController.cs]: won't work with liquid, maybe you can cache the address instead of the ScriptPubKey ?

**Comment by NicolasDorier** [comment]
This need rebase, this is lot's of change for a small feature though.
What about removing the `WalletReceiveCacheUpdater.cs`, and instead just having `WalletReceiveStateService.Get` an async method, which reserve the next address to NBXplorer and cache it for the next request in memory.

File: BTCPayServer/Controllers/WalletsController.cs (hunk 2 above)
Comment (Kukks) [reply BTCPayServer/Controllers/WalletsController.cs]: yeah this was before we added the address in nbx . Fixed

**Comment by Kukks** [comment]
Rebased

**Comment by Kukks** [comment]
> 
> 
> This need rebase, this is lot's of change for a small feature though.
> What about removing the `WalletReceiveCacheUpdater.cs`, and instead just having `WalletReceiveStateService.Get` an async method, which reserve the next address to NBXplorer and cache it for the next request in memory.
That's what I do though. I reserve and cache an address in memory until it is either:
* unreserved manually
* the wallet changes
* a payment comes in 

I don't think it is bad at all

**Pull request merged**